### DISCUSSION TITLE
0.7.0

### DISCUSSION THREAD

**Discussion by kaikalii** [uiua-lang/uiua#314] [Announcements]
### Language
- An entire Uiua codebase is now compiled before it is executed, rather than compiling and executing line-by-line
- Add the [`memo`](https://uiua.org/docs/memo) modifier, which memoizes a function
- Add the [`comptime`](https://uiua.org/docs/comptime) modifier, which runs a function at compile time
- [`&i`](https://uiua.org/docs/&i) can now only be used as the first function in a binding
- [`repeat` `⍥`](https://uiua.org/docs/repeat) can no longer use a negative number of repetitions
- [`repeat` `⍥`](https://uiua.org/docs/repeat) can now be used with [`un` `°`](https://uiua.org/docs/un) and [`under` `⍜`](https://uiua.org/docs/under)
- [`reshape` `↯`](https://uiua.org/docs/reshape) now works with [`under` `⍜`](https://uiua.org/docs/under)
- [`scan` `\`](https://uiua.org/docs/scan) now works with [`un` `°`](https://uiua.org/docs/un) in some cases
- [`setinv`](https://uiua.org/docs/setinv) and [`setund`](https://uiua.org/docs/setund) are no longer experimental
- Add output comments, which the formatter fills with values from the stack
  - Make an empty comment starting with `n` additional `#`s
  - The formatter will replace the comment with the top `n` values from the stack
  - Output comments in functions will show a number of values present on the stack for each time the function is called
### Interpreter
- LSP improvements
  - Add hover information on binding references
  - Add signatures to binding hover information
  - Add same-file binding rename support
  - Add same-file goto definition support
- Add the `uiua build` command, which emits a `.uasm` bytecode file
- `uiua run` can now run a `.uasm` bytecode file
- `uiua stand` now embeds the bytecode assembly in the executable
- Multiple compiler errors can now be emitted at once
- Bug and crash fixes
- Performance improvements
### Website
- Add an [Inverses](https://uiua.org/docs/inverses) tutorial
- Each tutorial challenge now contains 1 or 2 answers

<hr /><em>This discussion was created from the release <a href='https://github.com/uiua-lang/uiua/releases/tag/0.7.0'>0.7.0</a>.</em>

**Comment by JoelNiemela** [comment]
Is there any documentation on the `uiua build` command?

Specifically, I'm wondering if `uiua build main.ua && uiua run main.uasm` is effectively the same as running `uiua run.ua`, or are some features not yet implemented/supported by `.uasm`?

I'm asking because `uiua --help` shows the message "Build an assembly (the .uasm format is currently unstable)" for the `build` command, and I'm unsure what "unstable" means in this context.

**Reply by kaikalii** [reply]
The command should be effectively the same. 

`.uasm` is not documented because it is very unstable and not well-tested. By unstable I mean the entire format could change. It has no spec and no guarantees.

**Reply by JoelNiemela** [reply]
Understood, thank you!
